File: app/src/main/java/com/spidroid/starry/activities/ReportActivity.java
package com.spidroid.starry.activities;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.spidroid.starry.R; // تأكد من أن هذا الاستيراد صحيح

import java.util.HashMap;
import java.util.Map;

public class ReportActivity extends AppCompatActivity {

    private static final String TAG = "ReportActivity";

    public static final String EXTRA_REPORTED_ITEM_ID = "REPORTED_ITEM_ID";
    public static final String EXTRA_REPORT_TYPE = "REPORT_TYPE"; // "post", "comment", "user"
    public static final String EXTRA_REPORTED_AUTHOR_ID = "REPORTED_AUTHOR_ID"; // معرّف صاحب المحتوى

    private FirebaseFirestore db;
    private FirebaseAuth auth;
    private FirebaseUser currentUser;

    private TextView tvReportingItemInfo;
    private RadioGroup rgReportReasons;
    private EditText etReportDetails;
    private Button btnSubmitReport;
    private ProgressBar pbLoading;

    private String reportedItemId;
    private String reportType;
    private String reportedAuthorId;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_report);

        db = FirebaseFirestore.getInstance();
        auth = FirebaseAuth.getInstance();
        currentUser = auth.getCurrentUser();

        if (currentUser == null) {
            Toast.makeText(this, "You need to be logged in to report content.", Toast.LENGTH_LONG).show();
            finish();
            return;
        }

        reportedItemId = getIntent().getStringExtra(EXTRA_REPORTED_ITEM_ID);
        reportType = getIntent().getStringExtra(EXTRA_REPORT_TYPE);
        reportedAuthorId = getIntent().getStringExtra(EXTRA_REPORTED_AUTHOR_ID); // استلام المعرّف

        if (TextUtils.isEmpty(reportedItemId) || TextUtils.isEmpty(reportType)) {
            Toast.makeText(this, "Invalid report data.", Toast.LENGTH_LONG).show();
            finish();
            return;
        }

        Toolbar toolbar = findViewById(R.id.toolbar_report);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setTitle("Report " + capitalize(reportType));
        }
        toolbar.setNavigationOnClickListener(v -> finish());

        tvReportingItemInfo = findViewById(R.id.tv_reporting_item_info);
        rgReportReasons = findViewById(R.id.rg_report_reasons);
        etReportDetails = findViewById(R.id.et_report_details);
        btnSubmitReport = findViewById(R.id.btn_submit_report);
        pbLoading = findViewById(R.id.pb_report_loading);

        tvReportingItemInfo.setText("You are reporting a " + reportType + " (ID: " + reportedItemId + ")");

        btnSubmitReport.setOnClickListener(v -> submitReport());
    }

    private String capitalize(String str) {
        if (str == null || str.isEmpty()) {
            return str;
        }
        return str.substring(0, 1).toUpperCase() + str.substring(1);
    }

    private void submitReport() {
        int selectedReasonId = rgReportReasons.getCheckedRadioButtonId();
        if (selectedReasonId == -1) {
            Toast.makeText(this, "Please select a reason for reporting.", Toast.LENGTH_SHORT).show();
            return;
        }

        RadioButton selectedRadioButton = findViewById(selectedReasonId);
        String reason = selectedRadioButton.getText().toString();
        String details = etReportDetails.getText().toString().trim();

        pbLoading.setVisibility(View.VISIBLE);
        btnSubmitReport.setEnabled(false);

        Map<String, Object> reportData = new HashMap<>();
        reportData.put("reportingUserId", currentUser.getUid());
        reportData.put("reportedItemId", reportedItemId);
        reportData.put("reportedItemType", reportType);
        if (reportedAuthorId != null && !reportedAuthorId.isEmpty()) { // إضافة reportedAuthorId إذا كان موجودًا
            reportData.put("reportedAuthorId", reportedAuthorId);
        }
        reportData.put("reason", reason);
        if (!details.isEmpty()) {
            reportData.put("details", details);
        }
        reportData.put("timestamp", FieldValue.serverTimestamp());
        reportData.put("status", "pending"); // "pending", "reviewed_accepted", "reviewed_rejected"

        db.collection("reports")
                .add(reportData)
                .addOnSuccessListener(documentReference -> {
                    pbLoading.setVisibility(View.GONE);
                    btnSubmitReport.setEnabled(true);
                    Toast.makeText(ReportActivity.this, "Report submitted successfully. Thank you.", Toast.LENGTH_LONG).show();
                    finish();
                })
                .addOnFailureListener(e -> {
                    pbLoading.setVisibility(View.GONE);
                    btnSubmitReport.setEnabled(true);
                    Log.e(TAG, "Error submitting report", e);
                    Toast.makeText(ReportActivity.this, "Failed to submit report: " + e.getMessage(), Toast.LENGTH_LONG).show();
                });
    }
}

File: app/src/main/java/com/spidroid/starry/activities/ReportActivity.kt
// hmze123/sambok/sambok-main/app/src/main/java/com/spidroid/starry/activities/ReportActivity.kt
package com.spidroid.starry.activities

import android.annotation.SuppressLint
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.RadioButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.spidroid.starry.R
import com.spidroid.starry.databinding.ActivityReportBinding
import java.util.Locale // ✨ تم إضافة هذا الاستيراد

class ReportActivity : AppCompatActivity() {

    private lateinit var binding: ActivityReportBinding
    private val db: FirebaseFirestore by lazy { Firebase.firestore }
    private val auth: FirebaseAuth by lazy { Firebase.auth }
    private var currentUser: FirebaseUser? = null

    private var reportedItemId: String? = null
    private var reportType: String? = null
    private var reportedAuthorId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityReportBinding.inflate(layoutInflater)
        setContentView(binding.root)

        currentUser = auth.currentUser
        if (currentUser == null) {
            Toast.makeText(this, "You must be logged in to report content.", Toast.LENGTH_LONG).show()
            finish()
            return
        }

        reportedItemId = intent.getStringExtra(EXTRA_REPORTED_ITEM_ID)
        reportType = intent.getStringExtra(EXTRA_REPORT_TYPE)
        reportedAuthorId = intent.getStringExtra(EXTRA_REPORTED_AUTHOR_ID)

        if (reportedItemId.isNullOrEmpty() || reportType.isNullOrEmpty()) {
            Toast.makeText(this, "Invalid report data provided.", Toast.LENGTH_LONG).show()
            finish()
            return
        }

        setupToolbar()
        setupUI()
    }

    private fun setupToolbar() {
        setSupportActionBar(binding.toolbarReport)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
            title = reportType?.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
        }
        binding.toolbarReport.setNavigationOnClickListener { finish() }
    }

    @SuppressLint("SetTextI18n")
    private fun setupUI() {
        binding.tvReportingItemInfo.text = "You are reporting a $reportType (ID: $reportedItemId)"
        binding.btnSubmitReport.setOnClickListener { submitReport() }
    }

    private fun submitReport() {
        val selectedReasonId = binding.rgReportReasons.checkedRadioButtonId
        if (selectedReasonId == -1) {
            Toast.makeText(this, "Please select a reason for reporting.", Toast.LENGTH_SHORT).show()
            return
        }

        val selectedRadioButton = findViewById<RadioButton>(selectedReasonId)
        val reason = selectedRadioButton.text.toString()
        val details = binding.etReportDetails.text.toString().trim()

        showLoading(true)

        val reportData = hashMapOf<String, Any?>(
            "reportingUserId" to (currentUser?.uid ?: "unknown"),
            "reportedItemId" to reportedItemId,
            "reportedItemType" to reportType,
            "reportedAuthorId" to reportedAuthorId, // Can be null if not provided
            "reason" to reason,
            "details" to if (details.isNotEmpty()) details else null,
            "timestamp" to FieldValue.serverTimestamp(),
            "status" to "pending"
        )

        db.collection("reports").add(reportData)
            .addOnSuccessListener {
                showLoading(false)
                Toast.makeText(this, "Report submitted successfully. Thank you.", Toast.LENGTH_LONG).show()
                finish()
            }
            .addOnFailureListener { e ->
                showLoading(false)
                Log.e(TAG, "Error submitting report", e)
                Toast.makeText(this, "Failed to submit report: ${e.message}", Toast.LENGTH_LONG).show()
            }
    }

    private fun showLoading(isLoading: Boolean) {
        binding.pbReportLoading.visibility = if (isLoading) View.VISIBLE else View.GONE
        binding.btnSubmitReport.isEnabled = !isLoading
    }

    companion object {
        private const val TAG = "ReportActivity"
        const val EXTRA_REPORTED_ITEM_ID = "REPORTED_ITEM_ID"
        const val EXTRA_REPORT_TYPE = "REPORT_TYPE"
        const val EXTRA_REPORTED_AUTHOR_ID = "REPORTED_AUTHOR_ID"
    }
}